### DISCUSSION TITLE
Supported data types

### DISCUSSION THREAD

**Discussion by nichtich** [pg-format/pg-formatter#14] [General]
Original PG had only string (and number?) types for property values. As proposed with https://github.com/g2glab/pg/issues/34 and [implemented in pgraph](https://github.com/gbv/pgraph/blob/main/src/pg.pegjs#L156-L211) this has been extended to scalar JSON types (string, number, boolean, null) because these types can easily be expressed in JSON.

However existing property graph database systems and languages support some other data types. See [this overview of scalar data types](https://docs.google.com/spreadsheets/d/1VO_lz7KbKOCJGQt08h-OS76BY-sxziL8FfnuwkKvEss/edit?usp=sharing) (not limited to property graphs). In particular `DATE`, `DATETIME`, `DURATION`, `POINT` are worth to consider.

Any extension beyond JSON types requires to extend PG format and PG-JSON, e.g.

    node :event date:2024-02-28 # parsed as date

or something like this, inspired by RDF/Turtle:

    node :event date:"2024-02-28"@DATE

and

~~~json
{
  "id": "node",
  "labels": ["event"],
  "properties": {
    "date": [
      { "type": "DATE", "value": "2024-02-28" }
    ]
  }
}
~~~

This seems to be a can of worms because there are many data types and many possible ways to encode them in JSON.

It should be noted that most property graph database systems don't allow mixed types in lists of property values. Both Kuzu and YARS-PG requires a schema to specify data type. In the example above this would be the information "property key `date` of node labeled `event` is a `DATE`. For this use case the data type can be omitted, so maybe there is no urgent need to support data types in PG format and PG-JSON.

**Comment by hchiba1** [comment]
Yeah, I prefer small specifications, so I've kept it simple (too simple until now!)

I understand problems and inconsistencies may not arise from our specification itself, but rather during the conversion to other formats, that is, in the interactions with other specifications.
Handling JSON is essential, and interaction with RDF Turtle could be the next candidate to be considered.

**Comment by nichtich** [comment]
There are three options:

a) Support JSON data types only, everything else must be handled on another level of abstraction
b) Add a fixed set of additional data types to PG
c) Introduce a mechanism to tag values with data types (like done [in RDF/Turtle](https://www.w3.org/TR/turtle/#turtle-literals) with `^^` or [in YAML](https://yaml.org/spec/1.2.2/#24-tags) with `!!`)

**Comment by hchiba1** [comment]
a) seems reasonable to me.

I made the minimum changes to obtain the desired behavior in this repository (294f2d4).
